compiler/rustc_codegen_ssa/src/mir/naked_asm.rs

```diff
@@ -0,0 +1,247 @@
+use crate::common;
+use crate::mir::FunctionCx;
+use crate::traits::{AsmMethods, BuilderMethods, GlobalAsmOperandRef, MiscMethods};
+use rustc_attr::InstructionSetAttr;
+use rustc_middle::bug;
+use rustc_middle::mir::mono::{Linkage, MonoItem, MonoItemData, Visibility};
+use rustc_middle::mir::InlineAsmOperand;
+use rustc_middle::ty;
+use rustc_middle::ty::layout::{HasTyCtxt, LayoutOf};
+use rustc_middle::ty::{Instance, TyCtxt};
+
+use rustc_span::sym;
+use rustc_target::asm::InlineAsmArch;
+
+impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
+    pub fn codegen_naked_asm(&self, instance: Instance<'tcx>) {
+        let cx = &self.cx;
+
+        let rustc_middle::mir::TerminatorKind::InlineAsm {
+            template,
+            ref operands,
+            options,
+            line_spans,
+            targets: _,
+            unwind: _,
+        } = self.mir.basic_blocks.iter().next().unwrap().terminator().kind
+        else {
+            bug!("#[naked] functions should always terminate with an asm! block")
+        };
+
+        let operands: Vec<_> =
+            operands.iter().map(|op| self.inline_to_global_operand(op)).collect();
+
+        let item_data = cx.codegen_unit().items().get(&MonoItem::Fn(instance)).unwrap();
+        let (begin, end) = crate::mir::naked_asm::prefix_and_suffix(cx.tcx(), instance, item_data);
+
+        let mut template_vec = Vec::new();
+        template_vec.push(rustc_ast::ast::InlineAsmTemplatePiece::String(begin));
+        template_vec.extend(template.iter().cloned());
+        template_vec.push(rustc_ast::ast::InlineAsmTemplatePiece::String(end));
```

Comment:
this performs a bunch of allocations. Not sure how big a deal that is.

Reply:
Duplicating the template seems unlikely to be a bottleneck anywhere. And you would probably have to touch areas of codegen used for stable `global_asm`, better not to worry about it unless this is shown to be a problem.